Index to data/new and safe rotate, build script cleanup

zas opened the pull request.

zas committed:
Introduce a safer way to do new/cur/old rotation

The script uses 3 directories named 0 1 2 and 3 symlinks named new old cur

new is renamed to cur atomically, and then new and old are re-linked.
There is no copy, new re-uses the (old-)old directory and clean it.

zas committed:
Ensure indexer creates new indexes in data/new and sanitize build script

- use new way to do the rotation
- get rid of the list of index types
- it fixes an issue when make fails, rotation should only happen if no error happened, we don't want to push bad or incomplete files (it happens ie. if memory lacks and indexer process is killed)
- only push if there are servers to push to, this allows to use the indexer on the same machine as the servlet
- ensure freedb processing timestamp is updated only if no error occured (this was the case for recording indexing)
- simplify push code

mayhem commented on bin/build:
```diff
@@ -62,8 +62,9 @@ if [ "$FREEDB_ELAPSED" -ge "$SEARCH_FREEDB_THRESHOLD" ] &&
-	run_and_show_errors make freedb
+	START=$(date +%s)
+	run_and_show_errors make freedb || exit 2
+	echo "$START" > $FREEDB_TIMESTAMP_FILE
```
Ahhh, good!

mayhem commented on bin/safe_dir_rotation:
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
```
Maybe add a usage statement or a couple of comments on how this should be invoked?

mayhem commented:
I think this makes sense to me. However, I am unclear what you are trying to solve here -- what problem does this fix?

zas commented:
[**@mayhem**](https://github.com/mayhem) commented on [5 déc. 2015 12:00 UTC+1](https://github.com/metabrainz/search-indexer/pull/1#issuecomment-162172651):

> I think this makes sense to me. However, I am unclear what you are trying to solve here -- what problem does this fix?

First, thanks for taking time to review this stuff.

Short explanation:
the whole process is not failure-resistant at all, we can't run indexer + search on the same machine easily with current scripts

Long explanation:

We have two way to rotate dirs:
- one is used in `build` script https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L110-L122
- one is used by https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/receive-from-server#L31 based on https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/smart-rotate.py

**So what i want to "fix":**
- https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/Makefile#L27 sets indexer write directory for new indexes to `./data/cur`, it would make much more sense to write to `./data/new`. Ie. if we want to dissociate index updates and index pushes (it could be two processes, one looping on index generation, the other syncing `./data/cur` after it was updated to a remote `./data/new`
- imagine indexer and servlet running on the same machine with current code: indexer writes to `./data/cur`, servlet reads from `./data/cur` -> boum !
- ensure potentially incomplete data is always written in `./data/new`, both on indexer (command from the `Makefile`), and the search (servlet) server (transfer using `netcat` + `tar`, in `receive-from-server`)
- when writing is ending successfully, do the rotation (new -> cur -> old). Note that current code fails on this, it potentially rotates even after a failure: 
  - if you have an error at https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L83 the script continue and execute https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L111 and following, rotating data
  - the same goes for https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L111 which may rotate after pushing bad data

This is work-in-progress, but i would improve that stuff to be more reliable, and simpler.

Another thing that annoys me: the push/receive stuff is using multiple commands, i don't think it is needed.
Push does: sync/rotate/restart and receive has a prepare which looks unused.

I think we could simplify: use scp to copy data using "normal" ssh key (disabling compression/encryption), and on receiver side, just check for a new file to detect the end of the process

Something like (not actual code):

On sender:
 touch END ; scp -r data/cur/\* receiver:~/data/new/ && scp END receiver:~/data/END

On receiver:
 while true; do
   if END change : rotate new -> cur && restart search-server
   sleep 
end

zas committed:
receive-from-server: use safe_dir_rotate

zas committed:
build: ensure data is set up

This should happen once only

zas committed:
build: on first run, set freedb/recording timestamps so indexes are built asap

- since files were created with the current time, it was preventing to build indexes of recordings or freedb just after installation
- code cleanup

zas committed:
build: give explicit reasons about why recording or freedb indexes are skipped

zas committed:
build: add a simple way to force recording or freedb builds

- create a file named "recording_build_force" or "freedb_build_force" in /home/search
- on next loop, it will be removed and the indexes will be built

zas committed:
build: use a file as flag, stopping indexing until new indexes are consumed

- when running on the same server, once new indexes are rotated to data/cur, the local search server has to be notified
- the file data/cur/UPDATE_READY can be copied over remote servers and used to detect the need to restart search server
- $SEARCH_HOME/is_search_server is another flag file, indicating the current server is also a search server

zas committed:
build: preserve skipped indexes, using hardlink copy

- skipped indexes were missing from the new directory
- copy using hardlinks, directories will be always on same partition

zas committed:
Do hardlink copy in safe_dir_rotation script

zas committed:
safe_dir_rotation: copy all missing indexes dirs from cur/ to new/

- if, for some reason, an index dir is missing, re-use previous one

zas committed:
safe_dir_rotation: copy_cur_new() returns cp exit code

zas committed:
build: if `pause_indexer` file exists, just sleep a bit and exists.

- it is an easy way to suspend indexer after a loop

zas committed:
build: improve pause message

zas committed:
build: make it possible to change pause delay using conf file

- SEARCH_INDEXER_PAUSE_DELAY

zas committed:
build: fix a typo, $SEARCH -> $SEARCH_HOME

zas committed:
build: cleanup and rename status/control files

recording_build_timestamp -> indexer/.recording_build_timestamp
freedb_build_timestamp -> indexer/.freedb_build_timestamp
Compat code provided for those.

New control files:
indexer/force_recording : force rebuilding recording indexes on next loop, autoremoved
indexer/force_freedb : force rebuilding freedb indexes on next loop, autoremoved
indexer/pause : pause the indexer, while the file exists

zas committed:
build: add a flag file to disable any push

zas committed:
Introduce SEARCH_SERVLET_PORT (defaults to 8080)

zas committed:
warm-up: fix up off-by-one error, add a run time limit, and randomize queries even more

zas committed:
warm-up: exit loop if too many http errors

zas committed:
warm-up: remove left over from debug

zas committed:
warm-up: use urllib.urlencode()

zas committed:
warm-up: count number of queries, display mean time at end

zas committed:
warm-up: allow one to pass duration as third parameter

zas committed:
warm-up: add queries/s

zas committed:
warm-up: declare t1/t0 outside the try/except block

zas committed:
warm-up: add headers, user-agent and cache (disabling)

zas committed:
Add a wrapper script for warm up, getting host and port from conf

zas committed:
do_warm_up: fix regex

zas committed:
push-to-server: check flag files to disable rotate or restart

zas committed:
build: always honor the delay, ignore return code of previous command

- the restart returns a failure, because of the wget

zas committed:
build: fix up typo

zas committed:
build: add log messages about sleep

zas committed:
restart_search_server: fix up svc method

- replace svc-jetty & svc-tomcat by just svc
- detect current server using ~search/is_search_server file
- use commands installed in /usr/local/sbin (allowed to run by sudoers)

zas committed:
Fix up runexclusive command path

/home/search/bin/freedb.cron: 12: exec: runexclusive: not found

yvanzo commented:
Closing as this repository has been replaced with [SIR](https://github.com/metabrainz/sir) in June 2018, see the [blog post](https://blog.metabrainz.org/2018/06/15/musicbrainz-search-overhaul/).
